### PR TITLE
Minor build improvements

### PR DESCRIPTION
**Pull request by rotu**
1. Don't skip `tsc` on install script. Previously, if `script/build.js` exists, we would skip the C++ build.
2. Fix issue with github tags if ZMQ version provided as semver. (github seems to normalize tarballed directory name by stripping leading `v` from tag)
3. Build scripts before node-gyp. Build library wrappers after.

Removed from this PR:
- ~~Remove questionable `.npmrc`~~
- ~~Switch to standard `node-gyp-build`. Modified version tends to hang on failure instead of crash.~~

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,9 +18,9 @@
   },
   "homepage": "http://zeromq.github.io/zeromq.js/",
   "dependencies": {
-    "@aminya/node-gyp-build": "4.5.0-aminya.5",
     "cross-env": "^7.0.3",
     "node-addon-api": "^7.0.0",
+    "node-gyp-build": "^4.5.0",
```
Comment (aminya) [inline package.json:23]: This change is out of date after #617

File: package.json [hunk 2]
```diff
@@ -79,20 +79,22 @@
...
     "build.script": "tsc -p ./script/tsconfig.json",
     "build.js": "run-p build.script build.library",
     "build.doc": "typedoc --options ./typedoc.json && minify-all -s docs-raw -d docs --jsCompressor terser",
     "deploy.doc": "run-s build.doc && gh-pages --dist \"./docs\"",
-    "prebuild": "run-s build.js && node ./script/prebuild.js",
+    "prebuild": "node ./script/prebuild.js",
     "build.native": "node-gyp configure --release && node-gyp build --release",
     "build.native.debug": "cross-env CMAKE_BUILD_TYPE=Debug node-gyp configure --debug && cross-env CMAKE_BUILD_TYPE=Debug node-gyp build --debug",
-    "build": "run-s build.js build.native",
-    "build.debug": "run-s build.js build.native.debug",
```
Comment (aminya) [inline package.json:95]: Why skipping the Js build before when it is needed?

File: script/build.ts [hunk 3]
```diff
@@ -7,9 +7,11 @@ const root = dirname(__dirname)
 function main() {
   const zmq_rev =
     // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing, @typescript-eslint/strict-boolean-expressions
-    process.env.ZMQ_VERSION || "20de92ac0a2b2b9a1869782a429df68f93c3625e"
-  const src_url = `https://github.com/zeromq/libzmq/archive/${zmq_rev}.tar.gz`
+    process.env.ZMQ_VERSION || "4.3.5"
 
+  // if it looks like a dot-separated version number, prepend with "v" to match repo tagging convention
+  const gitref = zmq_rev.match(/^\d+\./) ? `v${zmq_rev}` : zmq_rev
+  const src_url = `https://github.com/zeromq/libzmq/archive/${gitref}.tar.gz`
```
Comment (aminya) [inline script/build.ts:14]: I use an exact commit hash instead of tags because upstream is slow in tagging their releases.

**Review by aminya — CHANGES_REQUESTED** [review]

File: .npmrc [hunk 4]
```diff
@@ -1,6 +0,0 @@
-public-hoist-pattern[]=*
```
Comment (aminya) [inline .npmrc:1]: We want pnpm to act like npm so that we can guarantee successful build with npm

**Review by aminya — CHANGES_REQUESTED** [review]

**Review by aminya — CHANGES_REQUESTED** [review]
Please change the title of the pull request to follow the Code of Conduct. We want to have a friendly environment here in Zeromq
https://zeromq.org/community/code-of-conduct/

File: package.json (hunk 1 above)
Comment (rotu) [reply package.json:23]: Indeed it is!

File: package.json (hunk 2 above)
Comment (rotu) [reply package.json:95]: [The `preinstall` hook](https://docs.npmjs.com/cli/v9/using-npm/scripts#npm-install) is run before the `install` hook. That said, the scripts benefit very little from being written in TypeScript. I'd be in favor of converting them to plain JavaScript, or committing the compiled js to the repo so consumers can build the package from source *without* `typescript` installed.

File: script/build.ts (hunk 3 above)
Comment (rotu) [reply script/build.ts:14]: That makes a ton of sense and I love having the ability to have a hash. The version number is much more friendly. These changes still *allow* for a hash. I'll add a comment.

**Comment by rotu** [comment]
> Please change the title of the pull request to follow the Code of Conduct. We want to have a friendly environment here in Zeromq https://zeromq.org/community/code-of-conduct/

I'm sorry if my words or tone seemed like trolling or like a personal attack! I don't at all mean to antagonize you or anyone else who pours selfless hours into open source software. I intended it to be levity over how frustrating it is to get a stable build process and an allusion to the frustrated confusion in recent issues.

Thank you for the tone reminder. I'll try to be more constructive with my wording in the future!

File: .npmrc (hunk 4 above)
Comment (rotu) [reply .npmrc]: Okay. I'm not sure how this either makes pnpm act more like npm or makes npm act more like pnpm.

Pnpm discourages the use of `public-hoist-pattern` https://pnpm.io/npmrc#public-hoist-pattern, hence my suspicion:
> This setting is useful when dealing with some flawed pluggable tools that don't resolve dependencies properly.

Also, while not having two flavors of lockfile committed to this repo, I believe that there can be problems with various `npm` tooling if the lockfile doesn't exist at all.

**Review by aminya — CHANGES_REQUESTED** [review]
The building of JavaScript files doesn't need to happen on the user's machine since they are already included in the package. See #635

File: script/build.ts [hunk 5]
```diff
@@ -5,11 +5,13 @@ import {mkdir, cd, exec, find, mv} from "shelljs"
...
-    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing, @typescript-eslint/strict-boolean-expressions
-    process.env.ZMQ_VERSION || "5657b4586f24ec433930e8ece02ddba7afcf0fe0"
-  const src_url = `https://github.com/zeromq/libzmq/archive/${zmq_rev}.tar.gz`
+  // Revision to use. Can be a version number like "4.3.5", a branch name, or a commit hash.
+  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing, @typescript-eslint/strict-boolean-expressions
+  const zmq_rev = process.env.ZMQ_VERSION || "4.3.5"
 
+  // if it looks like a dot-separated version number, prepend with "v" to match repo tagging convention
+  const gitref = zmq_rev.match(/^\d+\./) ? `v${zmq_rev}` : zmq_rev
+  const src_url = `https://github.com/zeromq/libzmq/archive/${gitref}.tar.gz`
```
Comment (aminya) [inline script/build.ts:14]: Given the conflicts and recent changes, this can be a separate pull request, but it needs some testing to verify it works
